File: src/crewlet/a2a/service.py
"""A2AService — agent-to-agent channels over the durable queue.

Two things about this used to be process-local, and both of them were
load-bearing.

**The message content.** It rode an ``asyncio.Queue`` per channel while
the *wake* rode the seat's durable inbox. The target of an ask wakes on
the node that owns its seat, which is rarely the node that opened the
channel — so cross-node the wake arrived and the queue was empty, and
the agent was told *"they opened a channel but hasn't sent a message
yet"*. Content now rides the wake event itself, which is durable,
ordered per conversation, owner-routed and covered by the completion
ledger. There is no second delivery path to disagree with it.

**The channel's participants and state.** They were a dict, and every
authorization decision reads them: may this sender post, may this closer
close, who is the other party. Cross-node that dict is empty on the node
that has to answer, so a legitimate reply raised "channel is not open or
does not exist". They live in :mod:`crewlet.a2a.channels` now.

What did NOT need to become durable is the reply mechanism, because
there wasn't one. ``a2a_ask`` promises *"they reply on the same
channel"* and the wake prompt told the agent to call
``send_a2a_message`` — a tool that does not exist, and which
``tests/test_tools/test_builtin.py`` asserts is not registered. The
answering turn's final text is the reply now: no new LLM-facing tool, no
channel lifecycle for a model to manage, and the promise is kept.
"""

from __future__ import annotations

from typing import Any
from uuid import uuid4

from crewlet._logging import get_logger
from crewlet.a2a.channels import A2AChannelStore, MemoryA2AChannelStore
from crewlet.a2a.messages import ChannelMessage
from crewlet.events.types import (
    A2AChannelClosed,
    A2AChannelOpened,
    A2AMessageSent,
    Event,
)
from crewlet.queue.protocol import EventQueue
from crewlet.queue.topics import agent_inbox_topic

logger = get_logger("a2a.service")


def _trace_of(event: Any) -> dict[str, str]:
    """Trace context to stamp on an event caused by ``event``.

    ``Event``'s trace fields default to the AMBIENT span, which is the
    right answer everywhere a span is open — the ``a2a_ask`` tool runs
    inside the asker's Execute phase, so the ask inherits that trace for
    free. The reply does not: it is published from ``_answer_a2a``,
    which runs after ``run_turn`` has returned and its phase spans have
    closed, and the engine opens no span of its own. The ambient trace
    is then empty, and an event with no trace is unreachable from the
    exchange it belongs to — which is exactly what made an A2A event's
    trace link dead-end on the dashboard.

    So the causing event's context is copied forward explicitly, the
    same way ``A2AMessageDelivered`` already does it. Verbatim rather
    than re-parented: ``TurnEngine.run_turn`` calls ``restore_context``
    with these, so the woken turn becomes a child of the span that
    caused it, which is the true shape of the causal chain.

    Returns ``{}`` when there is nothing to copy, so the field defaults
    (ambient capture) still apply for callers that DO have a span.
    """
    trace_id = getattr(event, "trace_id", "") if event is not None else ""
    if not trace_id:
        return {}
    return {
        "trace_id": trace_id,
        "span_id": getattr(event, "span_id", "") or "",
        "parent_span_id": getattr(event, "parent_span_id", "") or "",
    }


class A2AService:
    """Manages A2A channel lifecycle within the Engine."""

    def __init__(
        self,
        queue: EventQueue,
        channels: A2AChannelStore | None = None,
    ) -> None:
        self._queue = queue
        self._channels: A2AChannelStore = channels or MemoryA2AChannelStore()
        self._handle_registry = None

    def set_channel_store(self, channels: A2AChannelStore) -> None:
        """Swap in the durable store once the database is known."""
        self._channels = channels

    @property
    def channels(self) -> A2AChannelStore:
        return self._channels

    def set_handle_registry(self, registry) -> None:
        """Wire the identity registry so the service can defend its
        own contract (the bus connects agent seats — see
        :meth:`request_channel`).  Set by the engine once the
        registry exists; ``None`` (tests, embedded use) skips the
        guard."""
        self._handle_registry = registry

    async def request_channel(
        self,
        requester: str,
        target: str,
        *,
        brief: str = "",
        sender_role: str = "",
        delegation_depth: int = 0,
        delegation_chain: list[str] | None = None,
        parent_turn_id: str = "",
    ) -> str:
        """Open a channel and wake the target, carrying the brief.

        The brief travels ON the wake event. It used to be pushed into a
        per-channel ``asyncio.Queue`` while only the wake was durable,
        which meant the content existed on exactly one node and the wake
        was delivered to whichever node owned the target's seat. Those
        are the same node only by luck.

        Raises ``ValueError`` when a registry is wired and ``target``
        is not an agent seat in the org.  This is the chokepoint that
        creates the inbox wake — without the guard, a human seat or a
        typo'd handle produces a channel whose wake event lands on a
        subscriber-less topic: the requester reports success and waits
        on a reply that can never come.  Callers (``a2a_ask``) surface
        the error as a tool failure so it stays visible — the agent
        re-plans, or reaches the target on a human surface (Slack /
        Jira) instead.

        The question the guard asks is "does this seat exist and does
        it have an inbox?", NOT "is it running in this process".  A
        colleague owned by another node is a perfectly good A2A target:
        the wake lands on its inbox and that node consumes it.  Asking
        the local pool instead would have made every cross-node ask
        fail as a typo — the more nodes, the fewer colleagues an agent
        appears to have.
        """
        if requester and requester == target:
            # A channel to yourself has no responder. ``_handle_a2a``
            # decides who answers by comparing the woken seat against
            # the channel's requester, so a self-channel wakes the
            # asker, is read as an incoming ANSWER, and is never
            # replied to — a turn spent on a question nobody was asked,
            # and a row the idle sweep eventually closes. The agent
            # wanted to think, not to ask; say so.
            raise ValueError(
                f"a2a target '{target}' is you — a channel has two "
                f"parties. Reason it through in this turn instead of "
                f"asking yourself."
            )
        if self._handle_registry is not None:
            party = self._handle_registry.resolve_party(target)
            if party is None or party.is_human:
                raise ValueError(
                    f"a2a target '{target}' is not an agent seat — the A2A "
                    f"bus connects agents only (human seats and unknown "
                    f"handles have no inbox)"
                )
        channel_id = f"a2a-{uuid4().hex[:12]}"
        participants = [requester, target]
        await self._channels.open(channel_id, requester=requester, target=target)

        # Announce the channel and record the brief BEFORE the wake.
        #
        # The wake is what runs the other agent's turn, and on the
        # in-process queue that happens inline: publishing it first put
        # the whole conversation — the answer, and the close — on the
        # observability topics ahead of the question that caused it, so
        # a dashboard trace read backwards. The order here is the order
        # things happened.
        opened_event = A2AChannelOpened(
            source=requester,
            channel_id=channel_id,
            requester=requester,
            target=target,
            participants=participants,
        )
        await self._queue.publish("crewlet.events.a2a_channel_opened", opened_event)

        if brief:
            await self._channels.count_message(channel_id)
            await self._publish_sent(channel_id, requester, target, brief, sender_role)

        chain = list(delegation_chain or [])
        if requester and requester not in chain:
            chain = [*chain, requester]
        wake_event = Event(
            type="a2a_request",
            source=requester,
            payload={
                "channel_id": channel_id,
                "requester": requester,
                "content": brief,
                "sender_role": sender_role,
            },
            delegation_depth=delegation_depth + 1,
            delegation_chain=chain,
            parent_turn_id=parent_turn_id,
        )
        await self._queue.publish(agent_inbox_topic(target), wake_event)

        logger.info(
            "channel_requested",
            channel_id=channel_id,
            requester=requester,
            target=target,
            participants=participants,
        )
        return channel_id

    async def reply(
        self,
        channel_id: str,
        sender: str,
        content: str,
        sender_role: str = "",
        *,
        question: str = "",
        caused_by: Any = None,
        delegation_depth: int = 0,
        delegation_chain: list[str] | None = None,
        parent_turn_id: str = "",
    ) -> None:
        """Answer on a channel and wake the other party.

        ``caused_by`` is the ask event this answers. Its trace context is
        copied onto the reply so the whole exchange — the ask, the
        answering turn's phases, the reply, and the turn it wakes — reads
        as ONE trace. Without it the reply is published outside any span
        (see :func:`_trace_of`) and carries no trace at all, which leaves
        an A2A event's trace link on the dashboard pointing nowhere.

        ``question`` echoes the original ask back to the asker. Its turn
        ended when it asked, and nothing rehydrates it — so without the
        echo the woken turn receives an answer with no record of what it
        asked, and has to reconstruct that from memory retrieval or act
        on the answer blind. This is the one wake path with no external
        surface to re-read, so the echo is the only way that context
        survives the round trip.

        The reply carries the ask's delegation depth UNCHANGED. The ask
        is the delegation; the answer is that same hop completing, and
        the cap (default 3) exists to bound how deep a chain of *asks*
        goes. Charging the return leg as well halves the budget in the
        one direction nobody meant to spend it: a scheduled 1:1 costs
        depth 1 to ask and 2 to answer, so the report's first follow-up
        question lands at 3 and the manager's turn dies on a
        ``guard_breach`` — a legitimate second exchange ending as an
        engine failure event. Counting only asks bounds the
        conversation at ``delegation_depth_limit`` exchanges, which is
        what the limit reads as.

        The CHAIN still grows on every hop: it is provenance, not a
        gate, and "alice → bob → alice" is exactly what happened.

        Raises:
            ValueError: the channel is unknown, already closed, or the
                sender is not a participant. All three used to be one
                silent drop; a reply that goes nowhere is the failure
                the requester experiences as "they never answered".
        """
        channel = await self._channels.get(channel_id)
        if channel is None:
            logger.error(
                "send_on_unknown_channel", channel_id=channel_id, sender=sender
            )
            raise ValueError(f"A2A channel '{channel_id}' does not exist")
        if not channel.is_open:
            logger.error("send_on_closed_channel", channel_id=channel_id, sender=sender)
            raise ValueError(f"A2A channel '{channel_id}' is closed")
        recipient = channel.other_party(sender)
        if not recipient:
            logger.error(
                "send_by_non_participant",
                channel_id=channel_id,
                sender=sender,
                participants=channel.participants,
            )
            raise ValueError(
                f"Sender '{sender}' is not a participant of channel '{channel_id}'"
            )

        trace = _trace_of(caused_by)
        await self._channels.count_message(channel_id)
        await self._publish_sent(
            channel_id, sender, recipient, content, sender_role, trace
        )

        chain = list(delegation_chain or [])
        if sender and sender not in chain:
            chain = [*chain, sender]
        wake_event = Event(
            type="a2a_message",
            source=sender,
            payload={
                "channel_id": channel_id,
                "sender": sender,
                "content": content,
                "sender_role": sender_role,
                "question": question,
            },
            delegation_depth=delegation_depth,
            delegation_chain=chain,
            parent_turn_id=parent_turn_id,
            **trace,
        )
        await self._queue.publish(agent_inbox_topic(recipient), wake_event)

        logger.info(
            "message_sent",
            channel_id=channel_id,
            sender=sender,
            sender_role=sender_role,
            recipient=recipient,
            content_length=len(content),
        )

    async def _publish_sent(
        self,
        channel_id: str,
        sender: str,
        recipient: str,
        content: str,
        sender_role: str,
        trace: dict[str, str] | None = None,
    ) -> None:
        msg = ChannelMessage(
            channel=channel_id,
            sender=sender,
            sender_role=sender_role,
            content=content,
        )
        await self._queue.publish(
            "crewlet.events.a2a_message_sent",
            A2AMessageSent(
                source=sender,
                channel_id=channel_id,
                sender=sender,
                sender_role=sender_role,
                recipient=recipient,
                message_id=str(msg.id),
                content=content,
                **(trace or {}),
            ),
        )

    async def close_channel(
        self, channel_id: str, closer: str = "", *, caused_by: Any = None
    ) -> None:
        """Close a channel and record it.

        Args:
            channel_id: The channel to close.
            closer: Handle of the agent requesting the close. When
                provided, must be a participant of the channel.
            caused_by: The event this close answers, if any. The close
                is published from the same traceless frame as the reply
                (see :func:`_trace_of`), so without it the event that
                ends the exchange falls outside the exchange's trace.

        Raises:
            ValueError: If *closer* is provided and is not a
                participant of the channel.
        """
        channel = await self._channels.get(channel_id)
        if closer and channel is not None and not channel.other_party(closer):
            logger.error(
                "close_by_non_participant",
                channel_id=channel_id,
                closer=closer,
                participants=channel.participants,
            )
            raise ValueError(
                f"'{closer}' is not a participant of channel '{channel_id}'"
            )

        closed = await self._channels.close(channel_id)
        if closed is None:
            # Already closed, or never existed. Closing is idempotent —
            # both sides may reasonably try — so this is not an error,
            # but it must not publish a second closed event either.
            logger.debug("channel_close_noop", channel_id=channel_id)
            return

        # Everything on the event comes from the row the close returned,
        # including the duration: both of its endpoints are then readings
        # of ONE clock (the database's, on the Postgres path) rather than
        # the opening node's and the closing node's opinions of the time.
        closed_event = A2AChannelClosed(
            source=closer or "system",
            channel_id=channel_id,
            closed_by=closer,
            participants=closed.participants,
            message_count=closed.message_count,
            duration_ms=closed.duration_ms,
            **_trace_of(caused_by),
        )
        await self._queue.publish("crewlet.events.a2a_channel_closed", closed_event)

        logger.info(
            "channel_closed",
            channel_id=channel_id,
            closed_by=closer or "system",
            participants=closed.participants,
            message_count=closed.message_count,
            duration_ms=round(closed.duration_ms),
        )

    async def close_idle_channels(self, older_than_seconds: float) -> int:
        """Close channels nothing has finished. Returns how many.

        A channel is closed by the answering turn, so the ones that
        reach here are the turns that never got to finish: a crash, or a
        node that died between the wake and the answer. Left open they
        are a row that never goes away and a promise to the requester
        that a reply may still arrive.
        """
        closed = await self._channels.close_idle(older_than_seconds)
        for channel in closed:
            # Named, counted and timed like any other close. These used
            # to publish with no participants, no count and a zero
            # duration — on precisely the channels worth tracing, since
            # a channel only reaches the idle sweep because the turn
            # that should have answered it never finished. "Who was
            # waiting on whom when that node died" was unanswerable from
            # the one event that exists to answer it.
            await self._queue.publish(
                "crewlet.events.a2a_channel_closed",
                A2AChannelClosed(
                    source="system",
                    channel_id=channel.channel_id,
                    closed_by="",
                    participants=channel.participants,
                    message_count=channel.message_count,
                    duration_ms=channel.duration_ms,
                ),
            )
        if closed:
            logger.info(
                "a2a_channels_closed_idle",
                count=len(closed),
                channels=[c.channel_id for c in closed],
            )
        return len(closed)

File: internal/a2a/service.go
package a2a

import (
	"context"
	"fmt"
	"slices"
	"time"

	"github.com/google/uuid"

	"github.com/crewlet/crewlet/internal/events"
	"github.com/crewlet/crewlet/internal/events/types"
	"github.com/crewlet/crewlet/internal/logging"
	"github.com/crewlet/crewlet/internal/queue"
	"github.com/crewlet/crewlet/internal/queue/topics"
	"github.com/crewlet/crewlet/internal/tracing"
)

var log = logging.Get("a2a")

// Directory answers where a handle's wake goes, and reports false when
// nowhere.
//
// The question is "does this seat exist and can it be woken", NOT "is it
// running in this process". A colleague owned by another node is a perfectly
// good target — the wake lands on its inbox and that node consumes it. Asking
// a local pool instead made every cross-node ask fail as a typo, so the more
// nodes a company ran, the fewer colleagues each agent appeared to have.
//
// ONE CALL ANSWERS BOTH HALVES, deliberately. This used to be IsAgentSeat,
// and the caller then built the wake's subject itself from the handle — so
// "is this seat addressable" and "what address is it" were two derivations
// that could disagree, which is exactly what they did the moment a seat's
// durable address stopped being its handle. A human seat, an unknown handle
// and a seat with no derivable id are one answer here, because a caller does
// the same thing with all three.
type Directory interface {
	SeatInbox(handle string) (uuid.UUID, bool)
}

// inbox is where a handle's wake goes, reporting false when nowhere.
func (s *Service) inbox(handle string) (uuid.UUID, bool) {
	return s.dir.SeatInbox(handle)
}

// Service opens channels and carries asks and answers over the durable queue.
type Service struct {
	channels Store
	queue    queue.Publisher
	dir      Directory

	// now is injectable so the suite can pin the clock. Nil takes
	// time.Now, so the zero-value path is the real one.
	now func() time.Time

	// newID mints channel ids. Injectable for the same reason.
	newID func() string
}

// Options configure a Service.
type Options struct {
	Directory Directory
	Now       func() time.Time
	NewID     func() string
}

// New builds a service.
func New(channels Store, pub queue.Publisher, opts Options) (*Service, error) {
	if channels == nil {
		return nil, fmt.Errorf("a2a: no channel store")
	}
	if pub == nil {
		return nil, fmt.Errorf("a2a: no publisher")
	}
	// REFUSED BY NAME rather than tolerated. The directory is what says
	// where a wake goes, so a service without one can open channels and
	// wake nobody — every ask succeeding and every answer never arriving,
	// which reads as a slow colleague rather than as a wiring mistake.
	if opts.Directory == nil {
		return nil, fmt.Errorf("a2a: no directory, so no ask could be " +
			"addressed — wire the running company's org")
	}
	s := &Service{channels: channels, queue: pub, dir: opts.Directory,
		now: opts.Now, newID: opts.NewID}
	if s.now == nil {
		s.now = func() time.Time { return time.Now().UTC() }
	}
	if s.newID == nil {
		s.newID = func() string { return "a2a-" + uuid.New().String()[:12] }
	}
	return s, nil
}

// Ask is one seat asking another.
type Ask struct {
	Requester string
	Target    string
	Brief     string

	// SenderRole is the requester's role name, so the answering prompt can
	// say who is asking rather than only which handle.
	SenderRole string

	DelegationDepth int
	DelegationChain []string

	// TurnID and WorkKey identify the ASKING turn — the run whose tool loop
	// called a2a_ask, and the unit of work it was dispatched for.
	//
	// ONE FIELD, TWO LANDINGS, and the coincidence is worth stating because
	// the names differ at each end. TurnID is stamped on the audit records
	// this call publishes, where it answers "what else happened on this
	// turn" for the Turn screen; it is ALSO copied onto the wake's envelope
	// as ParentTurnID, where it answers "what asked for this" for whoever
	// is woken. The caller is both those things at once — the turn writing
	// the record is the parent of the wake it triggers — so a second field
	// here would be one value spelled twice, with nothing keeping the two
	// equal.
	TurnID  string
	WorkKey string
}

// Open opens a channel and wakes the target, carrying the brief.
//
// The brief travels ON the wake event. It used to be pushed into a per-channel
// in-process queue while only the wake was durable, which meant the content
// existed on exactly one node and the wake was delivered to whichever node
// owned the target's seat — the same node only by luck.
func (s *Service) Open(ctx context.Context, ask Ask) (string, error) {
	if ask.Requester != "" && ask.Requester == ask.Target {
		// A channel to yourself has no responder: the answering side
		// decides who replies by comparing the woken seat against the
		// requester, so a self-channel wakes the asker, is read as an
		// incoming ANSWER, and is never replied to — a turn spent on a
		// question nobody was asked. The agent wanted to think, not to
		// ask; say so.
		return "", fmt.Errorf("%w: %q is you — reason it through in this turn "+
			"instead of asking yourself", ErrSelfChannel, ask.Target)
	}
	target, addressable := s.inbox(ask.Target)
	if !addressable {
		// The chokepoint that creates the inbox wake. Without the guard a
		// human seat or a typo'd handle produces a channel whose wake
		// lands on a subscriber-less topic: the requester reports success
		// and waits on a reply that can never come.
		return "", fmt.Errorf("%w: %q (human seats and unknown handles have "+
			"no inbox)", ErrNotAnAgent, ask.Target)
	}

	now := s.now()
	id := s.newID()
	ch := Channel{ID: id, Requester: ask.Requester, Target: ask.Target,
		OpenedAt: now, LastAt: now}
	if err := s.channels.Open(ctx, ch); err != nil {
		return "", err
	}

	// Announce the channel BEFORE the wake. The wake is what runs the
	// other agent's turn, and on an in-process queue that happens inline:
	// publishing it first put the whole conversation — the answer, and the
	// close — on the observability topics ahead of the question that caused
	// it, so a trace read backwards.
	opened := events.New(types.A2AChannelOpened{
		ChannelID: id, Requester: ask.Requester, Target: ask.Target,
		Participants: ch.Participants(),
		TurnID:       ask.TurnID, WorkKey: ask.WorkKey,
	}, tracing.TraceOf(ctx))
	opened.Source = ask.Requester
	if err := s.queue.Publish(ctx, topics.Event(opened.Type), opened); err != nil {
		return "", fmt.Errorf("a2a: announce channel %s: %w", id, err)
	}

	if ask.Brief != "" {
		if _, err := s.channels.CountMessage(ctx, id, now); err != nil {
			return "", err
		}
		if err := s.publishSent(ctx, types.A2AMessageSent{
			ChannelID: id, Sender: ask.Requester, Recipient: ask.Target,
			Content: ask.Brief, SenderRole: ask.SenderRole,
			TurnID: ask.TurnID, WorkKey: ask.WorkKey,
		}); err != nil {
			return "", err
		}
	}

	wake := events.New(types.A2ARequest{
		ChannelID: id, Requester: ask.Requester,
		SenderRole: ask.SenderRole, Content: ask.Brief,
	}, events.TraceContext{})
	wake.Timestamp = now
	wake.Source = ask.Requester
	// The ASK is the delegation, so this is the leg that charges.
	wake.DelegationDepth = ask.DelegationDepth + 1
	wake.DelegationChain = appendChain(ask.DelegationChain, ask.Requester)
	wake.ParentTurnID = ask.TurnID
	if err := s.queue.Publish(ctx, topics.AgentInbox(target), wake); err != nil {
		return "", fmt.Errorf("a2a: wake %s: %w", ask.Target, err)
	}

	log.InfoContext(ctx, "channel_requested", "channel_id", id,
		"requester", ask.Requester, "target", ask.Target)
	return id, nil
}

// Answer is one seat replying on an open channel.
type Answer struct {
	ChannelID  string
	Sender     string
	Content    string
	SenderRole string

	// Question echoes the original ask back to the asker.
	//
	// Its turn ENDED when it asked and nothing rehydrates it, so without
	// the echo the woken turn receives an answer with no record of what it
	// asked and must reconstruct that from memory or act on it blind. This
	// is the one wake path with no external surface to re-read, so the
	// echo is the only way that context survives the round trip.
	Question string

	// CausedBy is the ask this answers. Its trace context is copied onto
	// the reply so the whole exchange — the ask, the answering turn's
	// phases, the reply, and the turn it wakes — reads as ONE trace.
	CausedBy *events.Event

	DelegationDepth int
	DelegationChain []string

	// TurnID and WorkKey identify the ANSWERING turn — the run that
	// produced this reply. See [Ask.TurnID] for why one field feeds both
	// the audit record's turn id and the wake's parent pointer.
	TurnID  string
	WorkKey string
}

// Reply answers on a channel and wakes the other party.
//
// THE REPLY CARRIES THE ASK'S DELEGATION DEPTH UNCHANGED. The ask is the
// delegation; the answer is that same hop completing, and the cap exists to
// bound how deep a chain of ASKS goes. Charging the return leg halves the
// budget in the one direction nobody meant to spend it: a scheduled 1:1 costs
// depth 1 to ask and 2 to answer, so the report's first follow-up question
// lands at 3 and the manager's turn dies on a guard breach — a legitimate
// second exchange ending as an engine failure.
//
// The CHAIN still grows on every hop. It is provenance, not a gate, and
// "alice → bob → alice" is exactly what happened.
func (s *Service) Reply(ctx context.Context, ans Answer) error {
	ch, err := s.channels.Get(ctx, ans.ChannelID)
	if err != nil {
		return err
	}
	if !ch.Open() {
		return fmt.Errorf("%w: %s", ErrClosed, ans.ChannelID)
	}
	recipient, ok := ch.OtherParty(ans.Sender)
	if !ok {
		return fmt.Errorf("%w: %q is not on channel %s", ErrNotParticipant, ans.Sender, ans.ChannelID)
	}

	now := s.now()
	if _, err := s.channels.CountMessage(ctx, ans.ChannelID, now); err != nil {
		return err
	}
	if err := s.publishSent(ctx, types.A2AMessageSent{
		ChannelID: ans.ChannelID, Sender: ans.Sender, Recipient: recipient,
		Content: ans.Content, SenderRole: ans.SenderRole,
		TurnID: ans.TurnID, WorkKey: ans.WorkKey,
	}); err != nil {
		return err
	}

	wake := events.New(types.A2AMessage{
		ChannelID: ans.ChannelID, Sender: ans.Sender,
		SenderRole: ans.SenderRole, Question: ans.Question, Content: ans.Content,
	}, events.TraceContext{})
	wake.Timestamp = now
	wake.Source = ans.Sender
	wake.DelegationDepth = ans.DelegationDepth
	wake.DelegationChain = appendChain(ans.DelegationChain, ans.Sender)
	wake.ParentTurnID = ans.TurnID
	if ans.CausedBy != nil {
		wake.TraceID = ans.CausedBy.TraceID
		wake.SpanID = ans.CausedBy.SpanID
		wake.ParentSpanID = ans.CausedBy.ParentSpanID
	}
	seat, addressable := s.inbox(recipient)
	if !addressable {
		// A counterparty who has been removed or turned into a human seat
		// since the channel opened. The message is already recorded and
		// announced; what cannot happen is the wake.
		return fmt.Errorf("a2a: wake %s: %w (the counterparty has no inbox)",
			recipient, ErrNotAnAgent)
	}
	if err := s.queue.Publish(ctx, topics.AgentInbox(seat), wake); err != nil {
		return fmt.Errorf("a2a: wake %s: %w", recipient, err)
	}

	log.InfoContext(ctx, "message_sent", "channel_id", ans.ChannelID,
		"sender", ans.Sender, "recipient", recipient, "length", len(ans.Content))
	return nil
}

// Closure is one party closing a channel it is finished with.
//
// A STRUCT RATHER THAN FOUR STRINGS, matching [Ask] and [Answer], because the
// arguments are four values of one type and a positional call site gives a
// reader no way to catch a transposed pair. The alternative was to keep the
// identity on the Service or on the [Channel] record, and both are wrong in a
// way that is worth writing down: a Service is one long-lived struct shared by
// every seat this node runs, so a turn's id parked on it is a data race that
// attributes one seat's close to another seat's turn; and the Channel record
// carries the OPENING turn, which is the asker's, on a record whose whole
// subject is what the ANSWERING turn did with it.
type Closure struct {
	ChannelID string

	// ClosedBy is the participant closing it. Empty for a caller that is
	// not one — the maintenance sweep — which is what makes the "system"
	// reading of [types.A2AChannelClosed] honest.
	ClosedBy string

	// TurnID and WorkKey identify the closing turn. See [Ask.TurnID]; the
	// close publishes no wake, so this one lands in a single place.
	TurnID  string
	WorkKey string
}

// Close closes a channel and announces it.
//
// Closing an already-closed channel is not an error: both parties may close,
// and the second one is not a fault. The announcement is skipped in that case
// so a dashboard does not draw two closes for one channel.
func (s *Service) Close(ctx context.Context, c Closure) error {
	before, err := s.channels.Get(ctx, c.ChannelID)
	if err != nil {
		return err
	}
	if !before.Open() {
		return nil
	}
	now := s.now()
	ch, err := s.channels.Close(ctx, c.ChannelID, now)
	if err != nil {
		return err
	}
	return s.announceClose(ctx, ch, c, now)
}

// SweepIdle closes every channel idle since before cutoff and announces each
// one, reporting how many it closed.
//
// HERE RATHER THAN IN THE SWEEP JOB, because announcing is the reason
// [Store.CloseIdle] returns the channels it closed rather than a count — and
// the one caller discarded the list, so nothing was published at all. Three
// separate places described a close event the sweep never emitted: that
// method's own contract, the "system" fallback in
// [types.A2AChannelClosed.Summary], and the event-system doc. What an operator
// lost with it is the only signal that an ask went unanswered: the requester's
// turn ended when it asked, so a channel reaching this sweep means some turn
// never finished, and that is exactly the event worth seeing.
//
// The publisher lives in this package for the reason every other A2A publish
// does: what a close means on the wire is this package's decision, and
// internal/maintenance is a scheduler of jobs rather than a second author of
// event payloads.
//
// A FAILED ANNOUNCEMENT DOES NOT UNDO THE CLOSE, and does not stop the rest:
// the channel is already closed in the store, the sweep cannot roll that back,
// and abandoning the remaining channels would leave a batch half-reported with
// no record of where it stopped. The first error is returned once every
// channel has been attempted.
func (s *Service) SweepIdle(ctx context.Context, cutoff time.Time) (int, error) {
	now := s.now()
	closed, err := s.channels.CloseIdle(ctx, cutoff, now)
	if err != nil {
		return 0, err
	}
	var first error
	for _, ch := range closed {
		// No ClosedBy, no TurnID: a swept channel is one NO turn finished,
		// and naming this node's sweep as the closer would read as a
		// participant. See [Closure.ClosedBy].
		if err := s.announceClose(ctx, ch, Closure{ChannelID: ch.ID}, now); err != nil && first == nil {
			first = err
		}
	}
	return len(closed), first
}

// Purge deletes channels closed before cutoff, returning the count.
//
// A passthrough, so the retention sweep drives ONE surface rather than holding
// the store beside the service and choosing between them per job.
func (s *Service) Purge(ctx context.Context, cutoff time.Time) (int64, error) {
	return s.channels.Purge(ctx, cutoff)
}

// announceClose publishes the close record for an already-closed channel.
func (s *Service) announceClose(ctx context.Context, ch Channel, c Closure, now time.Time) error {
	ev := events.New(types.A2AChannelClosed{
		ChannelID: ch.ID, ClosedBy: c.ClosedBy,
		Participants: ch.Participants(),
		MessageCount: ch.Messages,
		// THE RECORD'S OWN TWO INSTANTS, never two machines' clocks: a
		// channel is opened on one node and closed on another as a matter
		// of course, so a duration taken across them would be skew.
		DurationMS: float64(ch.Duration(now).Milliseconds()),
		TurnID:     c.TurnID, WorkKey: c.WorkKey,
	}, tracing.TraceOf(ctx))
	ev.Source = c.ClosedBy
	if err := s.queue.Publish(ctx, topics.Event(ev.Type), ev); err != nil {
		return fmt.Errorf("a2a: announce close of %s: %w", ch.ID, err)
	}
	return nil
}

// publishSent records one message on a channel.
//
// It takes the PAYLOAD rather than the six strings it used to, because that is
// what the two call sites are assembling either way and six adjacent strings
// in a positional call is a transposition nothing catches — the compiler least
// of all.
func (s *Service) publishSent(ctx context.Context, sent types.A2AMessageSent) error {
	ev := events.New(sent, tracing.TraceOf(ctx))
	ev.Source = sent.Sender
	if err := s.queue.Publish(ctx, topics.Event(ev.Type), ev); err != nil {
		return fmt.Errorf("a2a: record message on %s: %w", sent.ChannelID, err)
	}
	return nil
}

// appendChain adds handle to the provenance chain if it is not already there.
//
// A copy, never an append in place: the caller's chain came off the triggering
// event and appending into its spare capacity would rewrite the record of a
// hop that already happened.
func appendChain(chain []string, handle string) []string {
	if handle == "" || slices.Contains(chain, handle) {
		return slices.Clone(chain)
	}
	return append(slices.Clone(chain), handle)
}
